powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/TemporaryKeyBehavior.java

```diff
@@ -0,0 +1,371 @@
+
+                final PrivateKey privateKey = keyConvertor.convertBytesToPrivateKey(Base64.getDecoder().decode(masterKeyPairEntity.getMasterKeyPrivateBase64()));
+                final PublicKey publicKey = keyConvertor.convertBytesToPublicKey(Base64.getDecoder().decode(masterKeyPairEntity.getMasterKeyPublicBase64()));
+
+                final byte[] secretKeyBytes = Base64.getDecoder().decode(applicationSecret);
+
+                final TemporaryKeyResult result = new TemporaryKeyResult();
+                result.setSecretKeyBytes(secretKeyBytes);
+                result.setPrivateKey(privateKey);
+                result.setPublicKey(publicKey);
+                return result;
+            } else {
+
+                final String appId = applicationVersionEntity.getApplication().getId();
+
+                final Optional<ActivationRecordEntity> activationWithoutLock = activationRepository.findActivationWithoutLock(requestClaims.getActivationId());
+                if (activationWithoutLock.isEmpty()) {
+                    throw new IllegalArgumentException("Activation ID not found.");
+                }
+                final ActivationRecordEntity activation = activationWithoutLock.get();
+                if (activation.getActivationStatus() != ActivationStatus.ACTIVE || !appId.equals(activation.getApplication().getId())) {
+                    throw new IllegalArgumentException("Activation ID not found.");
+                }
+
+                final EncryptionMode encryptionMode = activation.getServerPrivateKeyEncryption();
+                final String serverPrivateKeyBase64 = activation.getServerPrivateKeyBase64();
+                final ServerPrivateKey serverPrivateKeyEncrypted = new ServerPrivateKey(encryptionMode, serverPrivateKeyBase64);
+                String decryptedServerPrivateKey = serverPrivateKeyConverter.fromDBValue(serverPrivateKeyEncrypted, activation.getUserId(), activation.getActivationId());
+                final byte[] serverPrivateKeyBytes = Base64.getDecoder().decode(decryptedServerPrivateKey);
+                final PrivateKey serverPrivateKey = keyConvertor.convertBytesToPrivateKey(serverPrivateKeyBytes);
+
+                final byte[] serverPublicKeyBytes = Base64.getDecoder().decode(activation.getDevicePublicKeyBase64());
+                final PublicKey serverPublicKey = keyConvertor.convertBytesToPublicKey(serverPublicKeyBytes);
+
+                final byte[] devicePublicKeyBytes = Base64.getDecoder().decode(activation.getDevicePublicKeyBase64());
+                final PublicKey devicePublicKey = keyConvertor.convertBytesToPublicKey(devicePublicKeyBytes);
+                final SecretKey transportKey = keyFactory.deriveTransportKey(serverPrivateKey, devicePublicKey);
+                final byte[] transportKeyBytes = keyConvertor.convertSharedSecretKeyToBytes(transportKey);
```

Comment:
I suggested to use `HMAC_SHA256(key: KEY_TRANSPORT, data: APP_SECRET)` instead of direct usage of `KEY_TRANSPROT`. We're trying to avoid dual use of key in this PR, so we should not introduce an yet another example of dual use.

Reply:
We can adjust the key like you suggest...